includes/class-llms-order-generator.php

```diff
@@ -0,0 +1,733 @@
+
+	/**
+	 * User Action: validate and then commit (register or update) the user.
+	 *
+	 * @var string
+	 */
+	const UA_COMMIT = 'commit';
+
+	/**
+	 * User Action: perform user validation only.
+	 *
+	 * @var string
+	 */
+	const UA_VALIDATE = 'validate';
+
+	/**
+	 * The coupon used to discount the order.
+	 *
+	 * Derived from `$this->data['llms_coupon_code']`.
+	 *
+	 * Will be empty until the coupon is validated.
+	 *
+	 * @var LLMS_Coupon|null
+	 */
+	protected $coupon = null;
+
+	/**
+	 * Associative array of input data.
+	 *
+	 * Usually the $_POST superglobal.
+	 *
+	 * @var array
+	 */
+	protected $data = array();
+
+	/**
+	 * The payment gateway used to process the order.
+	 *
+	 * Derived from `$this->data['llms_payment_gateway']` .
+	 *
+	 * Will be empty until the gateway is validated.
+	 *
+	 * @var LLMS_Payment_Gateway|null
+	 */
+	protected $gateway = null;
+
+	/**
+	 * The access plan used to generate the order.
+	 *
+	 * Derived from `$this->data['llms_plan_id']`.
+	 *
+	 * Will be empty until the plan is validated.
+	 *
+	 * @var LLMS_Access_Plan|null
+	 */
+	protected $plan = null;
+
+	/**
+	 * The order.
+	 *
+	 * Derived from `$this->data['llms_order_key']`.
+	 *
+	 * Will be empty until the order is validated.
+	 *
+	 * This is only used during confirmation of existing orders.
+	 *
+	 * @var LLMS_Order|null
+	 */
+	protected $order = null;
+
+	/**
+	 * The student used to generate the order.
+	 *
+	 * Will be empty until the user is created / update following all validations.
+	 *
+	 * @var LLMS_Student|null
+	 */
+	protected $student = null;
+
+	/**
+	 * Constructor.
+	 *
+	 * @since [version]
+	 *
+	 * @param array $data {
+	 *     An associative array of input data used to generate the order, usually from $_POST.
+	 *
+	 *     @type integer $llms_plan_id         An LLMS_Access_Plan ID.
+	 *     @type string  $llms_agree_to_terms  A yes/no value determining whether or not the user has agreed to the site's terms.
+	 *     @type string  $llms_payment_gateway The ID of the payment gateway used to process the order.
+	 *     @type string  $llms_coupon_code     Optional. The coupon code string being used.
+	 *     @type string  $llms_order_key       Optional. An `LLMS_Order` key used to modify an existing pending order rather than creating a new one.
+	 *     @type array   ...$user_data         All remaining data is passed to the user creation functions.
+	 * }
+	 * @return void
+	 */
+	public function __construct( $data ) {
+		$this->data = $data;
+	}
+
+	/**
+	 * Confirms an existing pending order.
+	 *
+	 * @since [version]
+	 *
+	 * @return WP_Error|array Returns an array of data from the payment gateway's `confirm_pending_order()` method on success.
+	 */
+	public function confirm() {
+
+		$validate = $this->validate( true );
+		if ( is_wp_error( $validate ) ) {
+			return $validate;
+		}
+
+		$gateway_confirm = $this->gateway->confirm_pending_order( $this->order );
```

Comment:
Yes.

Moving forward gateways will choose how to implement `confirm_pending_order()`.

A gateway that implements AJAX functionality will return `WP_Error|array` while the "legacy" gateways (PayPal, Stripe, Auth.net, any other custom gateways) would never call this method and therefore returning `void()` would work just fine.

The only gateway that will utilize this will be the new PayPal gateway (in development): https://github.com/gocodebox/lifterlms-gateway-paypal/blob/34ff6938cea0de61e41f598777a32a1502e75167/includes/class-llms-payment-gateway-paypal-cp.php#L125-L137

Reply:
The return type of the abstract `LLMS_Payment_Gateway::confirm_pending_order()` should be changed from `void` to `mixed`.